### PR TITLE
Helion autotune and attention benchmarking

### PR DESCRIPTION
**Pull request by adelmuursepp**

### CODE REVIEW

File: python/workload_shapes.py [hunk 1]
```diff
@@ -43,4 +43,19 @@
     ProfilingTensor(('n', 16)), # B
     EmptyTensor(('m', 'n')),
     configs=GEMM_SHAPES
-)
\ No newline at end of file
+)
+
```
Comment (adelmuursepp) [inline python/workload_shapes.py]: I was able to run 16k sequence length initially because I did not need to run a torch reference kernel for it while autotuning. But when profiling I ran into OOM. Therefore capped the sequence length here

File: python/bench_kernels/helion/lora.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import torch
+import helion.language as hl
+
+
```
Comment (adelmuursepp) [inline python/bench_kernels/helion/lora.py]: Was to do test before

File: python/workload_shapes.py [hunk 3]
```diff
@@ -43,4 +43,27 @@
...
+]
+
+ATTENTION_ARGS = KernelArgs(
+    ProfilingTensor(('batch', 'seq_len', 'heads', 'head_dim')), # q (B, S, H, D)
+    ProfilingTensor(('batch', 'seq_len', 'heads', 'head_dim')), # k
+    ProfilingTensor(('batch', 'seq_len', 'heads', 'head_dim')), # v
+    configs=ATTENTION_SHAPES,
+)
+
+# The Helion kernel takes (B, H, S, D)
```
Comment (adelmuursepp) [inline python/workload_shapes.py]: Can't solve the ordering with a mask for attention args

File: python/workload_shapes.py (hunk 1 above)
Comment (dungwoong) [reply python/workload_shapes.py]: Sounds good, in my experiments I capped at 8k as well

File: python/workload_shapes.py (hunk 3 above)
Comment (dungwoong) [reply python/workload_shapes.py]: I have files for profiling attention decode and prefill, and they have a custom setup. I think as long as we're able to make the plots, the setup is fine so maybe we could go with that.

File: python/bench_kernels/helion/attention.py
Comment (dungwoong) [inline python/bench_kernels/helion/attention.py]: Would it be better to make a package in python like `helion_kernels` or something? So then eventually if we have other kernels we could just have packages like `helion_kernels` and then `some_other_kernels`

File: python/helion_utils/kernel_runner.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+def autotune_helion_kernel(kernel_fn, label, shapes_iter):
+    """
+    shapes is a callable that yields (key, [tensors]) tuples
+    """
+    for key, tensors in shapes_iter():
+        autotune_helion_kernel_single(kernel_fn, label, key, tensors)
+
+
+
+class HelionKernel:
```
Comment (dungwoong) [inline python/helion_utils/kernel_runner.py]: Wait it seems like other functions are using `autotune_helion_kernel_single` directly rather than using this class. Is that better for autotuning?

File: run_scripts/autotune_attention.sh [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)"
+
+# Default CSV with timestamp, specify CSV to overwrite it
+CSV="${1:-results/helion_attn_autotune_$(date +%Y%m%d_%H%M%S).csv}"
+
+export HELION_AUTOTUNE_COMPILE_TIMEOUT="${HELION_AUTOTUNE_COMPILE_TIMEOUT:-600}"
+export HELION_AUTOTUNE_PRECOMPILE="${HELION_AUTOTUNE_PRECOMPILE:-spawn}"
+# Helion uses os.cpu() to set HELION_AUTOTUNE_PRECOMPILE_JOBS default value
+# os.cpu() is 64 on slurm machine so use nproc to set the correct default value (specified by --cpus-per-task)
```
Comment (dungwoong) [inline run_scripts/autotune_attention.sh]: Could you add a sample command that you used for running this? We could also add a slurm script as a sample

File: python/bench_kernels/helion/attention.py
Comment (adelmuursepp) [reply python/bench_kernels/helion/attention.py]: Just a preference of naming. I followed python/kernels/[hel|c2|hl] style from c2. Personally it makes sense to have kernels as a separate folder separate from the other logic code/utils folders.

The other reason why it is useful to wrap in `bench_kernels` for now is that `cutedsl_kernels`, `triton_kernels` would shadow other submodules'.

Comment (adelmuursepp) [reply python/bench_kernels/helion/attention.py]: If you prefer to have separate kernel folders for different frameworks I can rewrite, let me know

File: python/helion_utils/kernel_runner.py (hunk 4 above)
Comment (adelmuursepp) [reply python/helion_utils/kernel_runner.py]: For autotuning I found it was much easier for debugging to know:
1. the autotuning time if it autotuned. or None for this if it was found in cache. This helped me see if there was something off like setting the correct precompile jobs to parallelize correctly
2. cache key written as label on the csv file to cross check

I didn't get that info from the class. But thinking about it now I can just edit the class. I think keeping that logic in class is a good idea.

File: python/workload_shapes.py (hunk 1 above)
Comment (adelmuursepp) [reply python/workload_shapes.py]: If it matters then I think it just happens to pytorch reference. So if you need it for your own kernels I think you can still include it

File: run_scripts/autotune_attention.sh (hunk 5 above)
Comment (adelmuursepp) [reply run_scripts/autotune_attention.sh]: Done

File: demo/helion/autotune_common.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    """
+    Autotunes for one shape using Helion 
+
+    output_cls decides the shape columns
+    """
+    tensors = prob_args.tensors(mask)
+
+    torch.cuda.synchronize()
+    t0 = time.perf_counter()
+    key, ran = runner.autotune(*tensors)
```
Comment (adelmuursepp) [inline demo/helion/autotune_common.py]: Tell me if this extra info while autotuning does not warrant a new method in the class and you want to rely on the compile method. I thought this extra info was useful for autotuning and timing will be useful to report for comparing to Helion

File: python/workload_shapes.py (hunk 3 above)
Comment (adelmuursepp) [reply python/workload_shapes.py]: Do you mean you want to remove the args for bhsd in this file?

Helion wants them in the bhsd order. Do you want me to transpose in the Helion attention file and remove from here?
